src/radical/pilot/pmgr/launching/default.py

```diff
@@ -835,13 +834,27 @@ def _start_pilot_bulk(self, resource, schema, pilots):
                 raise RuntimeError("SAGA Job state is FAILED. (%s)" % jd.name)
 
             pilot = None
-            pid   = jd.name
+
             for p in pilots:
-                if p['uid'] == pid:
+                # we do not force unique job_names and multiple pilots may have
+                # the same job_name. By checking if p['uid'] is in PMGR pilots
+                # we ensure that each pilot is checked only once.
```

Comment:
I wonder if we should actually require unique names.  I mean, otherwise a `cancel()` call can well end up on the wrong pilot if we did not get the order right in this piece of code?

Reply:
I checked with multiple pilots as well. It requires the name only during starting the pilots. After that it uses the job id that the scheduler provides to identify different jobs.

Reply:
Yes, sure - but I am not sure the mapping from pilot UID to pilot job is unique then.  For uniform pilots is should not matter, and we don't regularly run non-uniform pilots in the same session, so this is really an edge case.  And I don't have suggestion... :-P  So, ultimately this looks good to me - thanks!

Reply:
BTW just to confirm, isn't the order of jobs, jobs descriptions and pilots is consistent? (and kinda it should be forced if not). Thus the comparison of pilot and job names shouldn't be matter (and that extra loops could be skipped)
```python
for j, jd, pilot in zip(jc.get_tasks(), jd_list, pilots):
    ...
```

Reply:
@mtitov I am not sure I understand. Can you expand it a bit?

Reply:
@iparask yeah, that my proposal will allow to remove L835-853: list `jd_list` is formed from list `pilots` ([link](https://github.com/radical-cybertools/radical.pilot/blob/63be5c0fea8e4cae1dfdd8c10d84399a457fb191/src/radical/pilot/pmgr/launching/default.py#L683)), thus jobs descriptions order corresponds to the order of pilots, and we don't need to look for which pilot is related to which job description based on their names/uids, but rather use same index in their lists. And this [line](https://github.com/radical-cybertools/radical.pilot/blob/63be5c0fea8e4cae1dfdd8c10d84399a457fb191/src/radical/pilot/pmgr/launching/default.py#L828) could be converted into `for j, jd, pilot in zip(jc.get_tasks(), jd_list, pilots)`. Hope I didn't bring more confusion

Reply:
I see. This makes the assumption that the pilots are in the same order as their corresponding SAGA jobs. @andre-merzky is this true? If it is I think

Reply:
we didn't resolve this? btw, there already was an [assumption](https://github.com/radical-cybertools/radical.pilot/blob/63be5c0fea8e4cae1dfdd8c10d84399a457fb191/src/radical/pilot/pmgr/launching/default.py#L824) about SAGA jobs and job descriptions
```
        # we assume here that the tasks arrive in the same order as the job
        # descriptions.  For uniform sets of pilots the order does not matter
        # much though.  Either way, this needs confirming on SAGA level
        # FIXME
```
and list `jd_list` was created from list `pilots`, which assumes that the order is preserved. Just not dealing with pilot/job names (at least here) simplifies this part.